tests/database/test_header_db.py

```diff
@@ -242,3 +242,45 @@ def test_headerdb_canonical_header_retrieval_by_number(headerdb, genesis_header)
     for header in headers:
         actual = headerdb.get_canonical_block_header_by_number(header.block_number)
         assert_headers_eq(actual, header)
+
+
+def test_headerdb_header_exists(headerdb, genesis_header):
+    assert headerdb.header_exists(genesis_header.hash) is False
+    headerdb.persist_header(genesis_header)
+    assert headerdb.header_exists(genesis_header.hash) is True
+
+    chain_a = mk_header_chain(genesis_header, 3)
+    chain_b = mk_header_chain(genesis_header, 5)
+
+    assert not any(headerdb.header_exists(h.hash) for h in chain_a)
+    assert not any(headerdb.header_exists(h.hash) for h in chain_b)
+
+    for idx, header in enumerate(chain_a):
+        assert all(headerdb.header_exists(h.hash) for h in chain_a[:idx])
+        assert not any(headerdb.header_exists(h.hash) for h in chain_a[idx:])
+
+        # sanity pre-check
+        assert not headerdb.header_exists(header.hash)
```

Comment:
This seems over-cautious, since it is tested in 260 as well. Could be dropped.

Reply:
I'm inclined to leave it in because while they overlap, they test two distinctly different things.

- line 260 and 268) are testing properties about the chain of headers leading up to and following this header (and they happen to include this header in those checks)
- this line is checking that the specific header we are processing was correctly persisted.

Overhead should be mild so if you don't object....

Reply:
Eh, I guess not. When reading it, the duplication made me concerned that I misunderstood the earlier assertion and go back to re-read it. So it adds overhead to the read-through.

I guess a comment like `# rerun the existence test for this specific header` would have saved me the re-read.